automata/base/automaton.py

```diff
@@ -95,19 +108,20 @@ def _validate_final_states(self):
                     ', '.join(str(state) for state in invalid_states)))
 
     @property
-    def input_parameters(self):
+    def input_parameters(self) -> Dict[str, Any]:
         """Return the public attributes for this automaton."""
         return {attr_name: getattr(self, attr_name)
                 for attr_name in self.__slots__
                 if not attr_name.startswith('_')}
 
-    def copy(self):
+    #TODO add annotation to here?
```

Comment:
@eliotwrobson I would think a return type isn't needed here, but really depends on whether Python can infer the return type from the below call to `__init__` (this is something TypeScript does pretty well, but not sure how Python type-checkers handle this).

In other words, how good is mypy's or Pyright type inferencing ability? Considering that Pyright is developed by Microsoft (the same company that brought us TypeScript), I would think it would be pretty capable with inferencing.

Reply:
So with type annotations, the inferencing is actually pretty good in Pyright (which does a great job at powering Pylance, IMO), but the issue has to do with the fact that Python typechecking is very permissive by design. So even though Pylance will correctly infer `copy` returns the same type, other code that is typechecked won't strictly enforce this inference. 

Here's a concrete example: Without an explicit annotation, if client code calls the `copy` function and then called a method that did not exist, this would be incompatible with the inferred return type from `copy`, but it would be compatible with `Any` (as this doesn't contradict the existing blank annotation). This behavior also isn't uniform across typecheckers.

All of this is to say that Python typechecking is meant to be as flexible as possible, but this comes at the cost of being very permissive. I think it's generally a good idea to be as explicit as possible to get as much correctness out of the typechecker as we can.

Reply:
@eliotwrobson Well, as the Zen of Python states, "Explicit is better than implicit". This is the Way. 😉

So let's leave the TODO, and add the return type annotation for explicitness. I believe I should have access to commit to your branch, so I can update this now.